frameworks/D/kiss/source/app.d

```diff
@@ -0,0 +1,197 @@
+ *
+ * Developer: putao's Dlang team
+ *
+ * Licensed under the Apache-2.0 License.
+ *
+ */
+import std.stdio;
+
+import kiss.event;
+import kiss.net;
+import kiss.util.KissTimer;
+import kiss.util.thread;
+
+import std.conv;
+import std.json;
+import std.functional;
+import std.getopt;
+import std.exception;
+import std.experimental.logger;
+import std.datetime;
+import std.parallelism;
+import std.socket;
+import std.string;
+
+
+/**
+*/
+abstract class AbstractTcpServer
+{
+	protected EventLoopGroup _group = null;
+	protected bool _isStarted = false;
+	protected Address _address;
+
+	this(Address address, int thread = (totalCPUs - 1))
+	{
+		this._address = address;
+		_group = new EventLoopGroup(cast(uint) thread);
+	}
+
+	@property Address bindingAddress()
+	{
+		return _address;
+	}
+
+	void start()
+	{
+		if (_isStarted)
+			return;
+		debug writeln("start to listen:");
+
+		for (size_t i = 0; i < _group.length; ++i)
+		{
+			createServer(_group[i]);
+			debug writefln("lister[%d] created", i);
+		}
+		debug writefln("All the servers is listening on %s.", _address.toString());
+		_group.start();
+		_isStarted = true;
+	}
+
+	protected void createServer(EventLoop loop)
+	{
+		TcpListener listener = new TcpListener(loop, _address.addressFamily);
+
+		listener.reusePort(true);
+		listener.bind(_address).listen(1024);
+		listener.acceptHandler = &onConnectionAccepted;
+		listener.start();
+	}
+
+	protected void onConnectionAccepted(TcpListener sender, TcpStream client);
+
+	void stop()
+	{
+		if (!_isStarted)
+			return;
+		_isStarted = false;
+		_group.stop();
+	}
+}
+
+/**
+*/
+class HttpServer : AbstractTcpServer
+{
+	this(string ip, ushort port, int thread = (totalCPUs - 1))
+	{
+		super(new InternetAddress(ip, port), thread);
+	}
+
+	this(Address address, int thread = (totalCPUs - 1))
+	{
+		super(address, thread);
+	}
+
+	override protected void onConnectionAccepted(TcpListener sender, TcpStream client)
+	{
+		client.onDataReceived((in ubyte[] data) {
+			notifyDataReceived(client, data);
+		}).onClosed(() { notifyClientClosed(client); }).onError((string msg) {
+			writeln("Error: ", msg);
+		});
+	}
+
+	protected void notifyDataReceived(TcpStream client, in ubyte[] data)
+	{
+		debug writefln("on thread:%s, data received: %s", getTid(), cast(string) data);
+		string request = cast(string) data;
+		bool keepAlive = indexOf(request, " keep-alive", CaseSensitive.no) > 0;
+
+		ptrdiff_t index = indexOf(request, "/plaintext ", CaseSensitive.no);
+		if (index > 0)
+			respondPlaintext(client, keepAlive);
+		else if (indexOf(request, "/json ", CaseSensitive.no) > 0)
+		{
+			respondJson(client, keepAlive);
+		}
+		else
+		{
+			badRequest(client);
+		}
+
+	}
+
+	private void respondPlaintext(TcpStream client, bool keepAlive)
+	{
+		string writeData = "HTTP/1.1 200 OK\r\nContent-Length: 13\r\nConnection: Keep-Alive\r\nContent-Type: text/plain\r\nServer: Kiss/0.3\r\nDate: Wed, 17 Apr 2013 12:00:00 GMT\r\n\r\nHello, World!";
```

Comment:
This is a violation of the plaintext test requirements, which state:

> This test is not intended to exercise the allocation of memory or instantiation of objects. Therefore it is acceptable but not required to re-use a single buffer for the response text (`Hello, World`). However, the response must be fully composed from the response text and response headers within the scope of each request and **it is not acceptable to store the entire payload of the response, or an unnaturally large subset of the response, headers inclusive, as a pre-rendered buffer**. "Buffer" here refers to a byte array, byte buffer, character array, character buffer, string, or string-like data structure. The spirit of the test is to require the construction of the HTTP response as is typically done by a framework or platform via concatenation of strings or similar. For example, pre-rendering a buffer with `HTTP/1.1 200 OK<cr>Content-length: 15<cr>Server: Example<cr>` would not be acceptable.

To fix this, please make at least these changes:

1. Dynamically render the `Date` header value.
2. Dynamically render the `Content-Length` header value as you do already in the JSON test implementation.
3. Dynamically concatenate each HTTP header to the response separately.  For this one, I suppose it's sufficient to use a StringBuilder or whatever method of string concatenation is common in D code, just as long as the compiler isn't smart enough to optimize away the concatenation.